### PR TITLE
feat(application): send v2x messages in onShutdown

### PR DESCRIPTION
**Pull request by kschrab**
## Description

The main issue was, that vehicles could not send V2X messages within `onShutdown`. This was due to the fact, that all simulators (sns, cell, omnet++, ns-3) delete the vehicles from their simulation, if they appear in the `VehicleUpdates.removed` list. Since mosaic-application uses an event queue to call shutdown procedures of removed vehicles, it would _always_ be that shutdown is called _after_ the vehicle has already been removed from the communication simulator. Therefore there was no chance for V2X messages being sent.

In order to allow vehicles to send V2X message during their shutdown method, several issues have been addressed with this PR:

* In all simulators (cell, sns, Omnet++, ns-3) the `VehicleUpdates.removed` list will now be completely ignored. Instead, vehicle nodes are now removed whenever the communication for the vehicle should disabled (via `***CommunicationConfiguration`). Thanks to the existing architecture it is still possible to re-enable communication for vehicles, as the `latestVehicleUpdates` interaction is known in each of the simulator implementations.
* As a consequence, all vehicle units must disable their communication modules once they were shut down. To achieve this, during shutdown of `AbstractApplicationUnit`s in mosaic-application, both CellModule and AdHocModule are disabled if they were enabled before.
* mosaic-cell handles message transmissions using an internal event manager (`ChainManager`). However, configuration interactions were processed immediately with no respect to other events being in the queue. As consequence, if an application would call `sendMessage` and `disable()` on after another, the `sendMessage` call would be queued in the `ChainManager`, whereas `disable` would be executed instantly. To solve this, all processing of `CellModuleConfiguration` interactions are now scheduled by the `ChainManager` to preserve to correct order of execution.
* A new integration test `CellSnsSendMessageOnShutdownIT` lets five vehicles send a V2X message during their shutdown procedure via cell and adhoc topocast to an server respective rsu. It is checked if all messages are received.

During implementation, further issues came up:

* The shutdown of all other units (rsus, servers) were initiated from the `finishSimulation` call, which is already _outside_ the simulation loop (see `SequentialTimeManagement`). For example, this would mean that no interactions could be send within the `onShutdown` method. To solve this, mosaic-application now schedules an event at the `endTime` of the simulation at which all shutdown procedures are called. 
* In order to make the previous change work, the simulation loop in `SequentialTimeManagement` now includes the `endTime` as the last simulation step. (!) This change affects only ns3-federate, which needs to be adjusted accordingly (see https://github.com/mosaic-addons/ns3-federate/pull/9). 
* When using `EncodedPayload.EMPTY_PAYLOAD` or `V2xMessage.Empty` to send dummy messages, e.g. for testing, it was found that mosaic-cell does not support message size of 0 bytes. However, I fixed this in the `CapacityUtility` so that such message would just consume no bandwidth and can therefore be simulated.

Open Issues:
* In `AbstractNetworkAmbassador` (this is used for Omnet++ and ns-s3): It is currently not possible to _disable_ communication of nodes, they only can be removed. One reason for this is, that the `proto` file does not allow the `primaryRadioInterface `to be `null`. As a result, the simulation aborts as soon as a configuration with no radios is transmitted (which is currently used to mark a module configuration as "disable"). In the current solution, these kind of configurations are not sent anymore, instead all units are removed when an off-configuration is received. This works well for vehicles, but for static units this comes with the down side that they could never be re-enabled again. This might be addressed in a follow-up.

## Issue(s) related to this PR

* Resolves internal issue 563
  
## Affected parts of the online documentation

Changes in the documentation required?

No

## Definition of Done

**Prerequisites**

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] Your GitHub user id is linked with your Eclipse Account.

**Required**

- [x] The title of this merge request follows the scheme `type(scope): description`  (in the style of [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary))
- [x] You have assigned a suitable label to this pull request (e.g., `enhancement`, or `bugfix`)
- [x] `origin/main` has been merged into your Fork.
- [x] Coding guidelines have been followed (see CONTRIBUTING.md).
- [ ] All checks on GitHub pass.
- [ ] All tests on [Jenkins](https://ci.eclipse.org/mosaic/job/mosaic/) pass.

**Requested** (can be enforced by maintainers)

- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [x] If a bug has been fixed, a new unit test has been written (beforehand) to prove misbehavior
- [x] There are no new SpotBugs warnings.

### CODE REVIEW

File: fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java [hunk 1]
```diff
@@ -47,7 +45,7 @@ public final class CapacityUtility {
      * @throws IllegalArgumentException if delay or message size were set to invalid values
      */
     public static long calculateNeededCapacity(long messageSizeInBit, long delayInNs) throws IllegalArgumentException {
-        if (delayInNs <= 0 || messageSizeInBit <= 0) {
+        if (delayInNs <= 0 || messageSizeInBit < 0) {
```
Comment (schwepmo) [inline fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java:48]: Why did you change this?

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/ApplicationAmbassador.java [hunk 2]
```diff
@@ -226,6 +216,19 @@ public void initialize(final long startTime, final long endTime) throws Internal
...
+    }
+
+    private void shutdownSimulationUnits(Event event) {
+        SimulationKernel.SimulationKernel.setCurrentSimulationTime(event.getTime());
+
+        log.debug("remaining events: {}", eventScheduler.getAllEvents());
+        UnitSimulator.UnitSimulator.removeAllSimulationUnits();
+
+        log.debug("inform about finishSimulation");
+        SimulationKernel.SimulationKernel.finishSimulation();
```
Comment (schwepmo) [inline fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/ApplicationAmbassador.java:231]: is this method actually doing anything?

**Review by schwepmo — CHANGES_REQUESTED** [review]
Overall fine! I left some suggestions and questions.

**Review by schwepmo — APPROVED** [review]

**Review by schwepmo — CHANGES_REQUESTED** [review]
Accidentally approved my previous review.

File: test/scenarios/cell-sns-send-message-on-shutdown/cell/network.json [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        "downlink": {
+            "unicast": {
+                "delay": {
+                    "type": "GammaRandomDelay",
+                    "minDelay": "50 ms",
+                    "expDelay": "150 ms"
+                },
+                "transmission": {
+                    "lossProbability": 0.0,
+                    "retries": 2
```
[the remark this answers is not in the document]
Comment (kschrab) [reply test/scenarios/cell-sns-send-message-on-shutdown/cell/network.json:26]: I'll fix it to "maxRetries". For completeness I still tend to keep the field here in the config.

File: test/scenarios/cell-sns-send-message-on-shutdown/cell/network.json [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    "globalNetwork": {
+        "uplink": {
+            "delay": {
+                "type": "GammaRandomDelay",
+                "minDelay": "50 ms",
+                "expDelay": "150 ms"
+            },
+            "transmission": {
+                "lossProbability": 0.0,
+                "maxRetries": 2
```
[the remark this answers is not in the document]
Comment (kschrab) [reply test/scenarios/cell-sns-send-message-on-shutdown/cell/network.json:13]: In this specific test we probably won't configure anything larger than 0, but for documentation purposes I would keep the field (and also if we copy this file again sometime for another test).

File: fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java (hunk 1 above)
Comment (kschrab) [reply fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java:48]: See description in PR:

> When using EncodedPayload.EMPTY_PAYLOAD or V2xMessage.Empty to send dummy messages, e.g. for testing, it was found that mosaic-cell does not support message size of 0 bytes. However, I fixed this in the CapacityUtility so that such message would just consume no bandwidth and can therefore be simulated.

Comment (schwepmo) [reply fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java:48]: Do we want to allow empty messages?

File: test/scenarios/cell-sns-send-message-on-shutdown/cell/network.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (schwepmo) [reply test/scenarios/cell-sns-send-message-on-shutdown/cell/network.json:13]: Alright, then we leave it this way!

**Review by schwepmo — APPROVED** [review]
Looks good to me!

File: fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java (hunk 1 above)
Comment (kschrab) [reply fed/mosaic-cell/src/main/java/org/eclipse/mosaic/fed/cell/utility/CapacityUtility.java]: We decided to revert this change. CELL again rejects empty messages. The test now sends an 8-byte message. We will evaluate in a follow-up issue how we can cope we empty payloads (e.g., by modelling additional header size added to all messages send by CELL or SNS)

**Review by schwepmo — APPROVED** [review]
